File: src/main/java/org/kite9/diagram/dom/painter/AbstractPainter.java
package org.kite9.diagram.dom.painter;

import java.util.Arrays;

import org.apache.commons.math.fraction.BigFraction;
import org.kite9.diagram.common.objects.OPair;
import org.kite9.diagram.common.objects.Pair;
import org.kite9.diagram.dom.elements.Kite9XMLElement;
import org.kite9.diagram.dom.elements.StyledKite9XMLElement;
import org.kite9.diagram.dom.model.HasSVGRepresentation;
import org.kite9.diagram.dom.processors.XMLProcessor;
import org.kite9.diagram.model.AlignedRectangular;
import org.kite9.diagram.model.Connected;
import org.kite9.diagram.model.Connection;
import org.kite9.diagram.model.Container;
import org.kite9.diagram.model.Decal;
import org.kite9.diagram.model.Diagram;
import org.kite9.diagram.model.DiagramElement;
import org.kite9.diagram.model.Label;
import org.kite9.diagram.model.Rectangular;
import org.kite9.diagram.model.Temporary;
import org.kite9.diagram.model.Terminator;
import org.kite9.diagram.model.position.Layout;
import org.kite9.diagram.model.position.RectangleRenderingInformation;
import org.kite9.framework.common.Kite9XMLProcessingException;
import org.w3c.dom.Attr;
import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.NamedNodeMap;


/**
 * Base class for painter implementations
 * 
 * @author robmoffat
 *
 * @param <X>
 */
public abstract class AbstractPainter implements Painter {

	protected DiagramElement r;

	@Override
	public void setDiagramElement(DiagramElement de) {
		this.r = de;
	}

	protected void addAttributes(StyledKite9XMLElement toUse, Element out) {
		copyAttributes(toUse, out);
		
		String id = r.getID();
		if (id.length() > 0) {
			out.setAttribute("id", id);
		}
		
		out.setAttribute("k9-elem", toUse.getTagName());
		out.setAttribute("class", toUse.getAttribute("class")+" kite9-"+toUse.getTagName());

		addInfoAttributes(out);
	}
	
	public void copyAttributes(Element from, Element to) {
        NamedNodeMap attributes = from.getAttributes();
        for (int i = 0; i < attributes.getLength(); i++)
        {
            Attr node = (Attr) attributes.item(i);
            if (!node.getName().equals("xmlns")) {
	            Attr copy = to.getOwnerDocument().createAttribute(node.getName());
	            copy.setNodeValue(node.getNodeValue());
	            to.setAttributeNode(copy);
            }
        }
    }

	protected void addInfoAttributes(Element out) {
		StringBuilder debug = new StringBuilder();
		if (r instanceof Rectangular) {
			debug.append("position: "+ ((Rectangular) r).getContainerPosition()+"; ");
		} 
		if (r instanceof AlignedRectangular) {
			debug.append("horiz: "+ ((AlignedRectangular) r).getHorizontalAlignment()+"; ");
			debug.append("vert: "+ ((AlignedRectangular) r).getVerticalAlignment()+"; ");
		}
		if (r instanceof Container) {
			debug.append("sizing: " + ((Container) r).getSizing() + "; ");
			debug.append("layout: " + ((Container) r).getLayout() + "; ");
			if (((Container)r).getLayout() == Layout.GRID) {
				RectangleRenderingInformation rri = (RectangleRenderingInformation) r.getRenderingInformation();
				debug.append("grid-size: ["+rri.gridXSize()+", "+rri.gridYSize()+"]; ");
				debug.append("cell-xs: ["+commaIntList(rri.getCellXPositions())+"]; ");
				debug.append("cell-ys: ["+commaIntList(rri.getCellYPositions())+"]; ");
			}
		}
		
		if (r instanceof Connected) {
			RectangleRenderingInformation rri = (RectangleRenderingInformation) r.getRenderingInformation();
			if (r.getParent() instanceof Container) {
				Container parent = (Container) r.getParent();
				RectangleRenderingInformation prri = parent.getRenderingInformation();
				Layout l = parent.getLayout();
				if (l == Layout.GRID) {
					Pair<Integer> scaledX = scale(rri.gridXPosition(), prri.gridXSize());
					Pair<Integer> scaledY = scale(rri.gridYPosition(), prri.gridYSize());
					debug.append("grid-x: "+scaledX+"; "); 
					debug.append("grid-y: "+scaledY+"; "); 
				}
			}
		}
		
		
		if (r instanceof Terminator) {
			Connection link = ((Terminator) r).getConnection();
			boolean from = link.getDecorationForEnd(link.getFrom()) == r;
			debug.append("terminates: "+link.getID()+"; ");
			debug.append("terminates-at: "+(from ? link.getFrom().getID() : link.getTo().getID())+"; ");
			debug.append("end: "+(from ? "from; " : "to; ")); 
		}
		
		if (r instanceof Connection) {
			Connection link = ((Connection) r);
			debug.append("link: ['"+link.getFrom().getID()+"','"+link.getTo().getID()+"']; ");
			debug.append("direction: "+((Connection)r).getDrawDirection()+"; ");
			if (((Connection)r).getRenderingInformation().isContradicting()) {
				debug.append("contradicting: yes; ");
			}
		}
		
		if (r instanceof Rectangular) {
			String usage = getUsage((Rectangular) r);
			debug.append("rectangular: "+usage+"; ");
		}
			
		if (r instanceof Rectangular) {
			RectangleRenderingInformation rri = ((Rectangular) r).getRenderingInformation();
			debug.append("rect-pos: "+rri.getPosition()+"; ");
			debug.append("rect-size: "+rri.getSize()+ "; "); 
		}
		
		
		out.setAttribute("k9-info", debug.toString());
	}

	private String commaIntList(double[] p) {
		return Arrays.stream(p).mapToObj(d -> "" + ((int) d)).reduce((a, b) -> a+", "+b).orElse("");
	}

	private Pair<Integer> scale(OPair<BigFraction> p, int s) {
		BigFraction a = p.getA();
		BigFraction as = a.multiply(s);
		BigFraction b = p.getB();
		BigFraction bs = b.multiply(s);
		return new Pair<Integer>(as.intValue(), bs.intValue());
	}

	private String getUsage(Rectangular rect) {
		if (rect instanceof Diagram) {			
			return "diagram";
		}else if (rect instanceof Decal) {
			return "decal";
		} else if (rect instanceof Connected) {
			return "connected";
		} else if (rect instanceof Label) {
			return "label";
		} else if (rect instanceof Terminator) {
			return "terminator";
		} else {
			return "unknown";
		}
	}

	/**
	 * Outputs any SVG-renderable temporary elements to the output.
	 */
	protected void handleTemporaryElements(Element out, Document d, XMLProcessor postProcessor) {
		if (r instanceof Container) {
			((Container) r).getContents().stream()
				.filter(c -> c instanceof Temporary)
				.filter(c -> c instanceof HasSVGRepresentation) 
				.forEach(c -> {
					Element e = ((HasSVGRepresentation)c).output(d, postProcessor);
					if (e != null) {
						e.setAttribute("k9-elem", "--temporary");
						e.setAttribute("id", c.getID());
						out.appendChild(e);
					}
				});
		}
	}
	

	protected void ensureNoChildKite9Elements(Element e) {
		if (e instanceof Kite9XMLElement) {
			if (((Kite9XMLElement) e).iterator().hasNext()) {
				throw new Kite9XMLProcessingException(e+" shouldn't have nested Kite9 elements - it's supposed to be a leaf (svg elements only). ", e);
			}
		} else {
			throw new Kite9XMLProcessingException("How is "+e+" not a Kite9 element? ", e);
		}
	}
}

File: visualization/kite9-visualization-common/src/main/kotlin/org/kite9/diagram/dom/painter/AbstractPainter.kt
package org.kite9.diagram.dom.painter

import org.kite9.diagram.common.fraction.LongFraction
import org.kite9.diagram.common.objects.OPair
import org.kite9.diagram.common.objects.Pair
import org.kite9.diagram.dom.model.HasSVGRepresentation
import org.kite9.diagram.dom.processors.XMLProcessor
import org.kite9.diagram.model.*
import org.kite9.diagram.model.position.Dimension2D
import org.kite9.diagram.model.position.Direction
import org.kite9.diagram.model.position.Layout
import org.w3c.dom.Document
import org.w3c.dom.Element

/**
 * Base class for painter implementations
 *
 * @author robmoffat
 */
abstract class AbstractPainter : Painter {

	protected var r: DiagramElement? = null

    override fun setDiagramElement(de: DiagramElement) {
        r = de
    }

    abstract fun shortPainterName() : String

    protected fun addInfoAttributes(out: Element) {
        val debug = StringBuilder()

        if (r is Port) {
            val pr = r as Port
            val rri = pr.getRenderingInformation()
            debug.append("port: true; ")
            debug.append("direction: "+lowercase(pr.getPortDirection())+"; ");
            debug.append("pos: " + xy(rri.position) + "; ")
        }

        if (r is SizedRectangular) {
            val sr = r as SizedRectangular
            debug.append(
                "margin: [" + sr.getMargin(Direction.UP) + " " + sr.getMargin(Direction.RIGHT) + " "
                        + sr.getMargin(Direction.DOWN) + " " + sr.getMargin(Direction.LEFT) + "]; "
            )
            debug.append(
                "padding: [" + sr.getPadding(Direction.UP) + " " + sr.getPadding(Direction.RIGHT) + " "
                        + sr.getPadding(Direction.DOWN) + " " + sr.getPadding(Direction.LEFT) + "]; "
            )
            debug.append(
                "min-size: [" + sr.getMinimumSize().width() + " "+sr.getMinimumSize().height()+ "]; "
            )
            debug.append(
                "sizing: [" + lowercase((r as SizedRectangular).getSizing(false)) + ", " +
                        lowercase((r as SizedRectangular).getSizing(true)) + "]; "
            )
        }
        if (r is AlignedRectangular) {
            debug.append("horiz: " + lowercase((r as AlignedRectangular).getHorizontalAlignment()) + "; ")
            debug.append("vert: " + lowercase((r as AlignedRectangular).getVerticalAlignment()) + "; ")
        }
        if (r is Container) {
            val c = r as Container
            debug.append("layout: " + lowercase((r as Container).getLayout()) + "; ")
            if (c.getLayout() === Layout.GRID) {
                val rri = c.getRenderingInformation()
                debug.append("grid-size: [" + rri.gridXSize() + ", " + rri.gridYSize() + "]; ")
                debug.append("cell-xs: [" + commaIntList(rri.cellXPositions) + "]; ")
                debug.append("cell-ys: [" + commaIntList(rri.cellYPositions) + "]; ")
            }
        }
        if (r is Rectangular) {
            val c = r as Rectangular
            val rri = c.getRenderingInformation()
            val usage = getUsage(r as Rectangular)
            debug.append("rectangular: $usage; ")
            debug.append("rect-pos: " + xy(rri.position) + "; ")
            debug.append("rect-size: " + xy(rri.size) + "; ")
            debug.append("position: " + (r as Rectangular).getContainerPosition() + "; ")

            if (c.getParent() is Container) {
                val parent = c.getParent() as Container?
                val prri = parent!!.getRenderingInformation()
                val l = parent.getLayout()
                if (l === Layout.GRID) {
                    val scaledX = scale(rri.gridXPosition(), prri.gridXSize())
                    val scaledY = scale(rri.gridYPosition(), prri.gridYSize())
                    debug.append("grid-x: $scaledX; ")
                    debug.append("grid-y: $scaledY; ")
                }
            }
        }
        if (r is Terminator) {
            val link = (r as Terminator).getConnection()
            val from = link.getDecorationForEnd(link.getFrom()) === r
            val direction = if (link.getDrawDirection() != null)
                if (from) Direction.reverse(link.getDrawDirection()) else link.getDrawDirection()
            else
                (r as Terminator).getArrivalSide()

            debug.append("terminates: " + link.getID() + "; ")
            debug.append("terminates-at: " + (if (from) link.getFrom().getID() else link.getTo().getID()) + "; ")
            debug.append("end: " + if (from) "from; " else "to; ")
            debug.append("direction: "+lowercase(direction)+"; ")
        }
        if (r is Connection) {
            val link = r as Connection
            debug.append("link: ['" + link.getFrom().getID() + "','" + link.getTo().getID() + "']; ")
            debug.append("direction: " + lowercase((r as Connection).getDrawDirection()) + "; ")
            if ((r as Connection).getRenderingInformation().isContradicting) {
                debug.append("contradicting: yes; ")
            }
        }

        if (r is Temporary) {
            debug.append("temporary: true; ")
        }

        if (r is Label) {
            val labelPlacement = (r as Label).getLabelPlacement()
            if (labelPlacement!=null) {
                debug.append("placement: " + lowercase(labelPlacement) + ";")
            }
            val end = (r as Label).getEnd()
            if (end != null) {
                debug.append("end: " + lowercase(end) + ";")
            }
        }

        debug.append("painter: "+this.shortPainterName()+"; ")

        out.setAttribute("k9-info", debug.toString())
    }

    fun lowercase(e: Any?): String {
        return e?.toString()?.lowercase() ?: "null";
    }

    private fun commaIntList(p: DoubleArray?): String {
        if (p == null) {
            return "";
        }
        return p
            .map { it.toInt().toString() }
            .fold("") { a, b -> "$a, $b"}
    }

    private fun scale(p: OPair<LongFraction>?, s: Int): Pair<Int> {
        val a = p!!.a
        val `as` = a.multiply(s)
        val b = p.b
        val bs = b.multiply(s)
        return Pair(`as`.intValue(), bs.intValue())
    }

    private fun xy(d: Dimension2D?) : String {
        return if (d == null) "" else "[" + d.x() + ", "+d.y()+"]"
    }

    private fun getUsage(rect: Rectangular): String {
        return if (rect is Diagram) {
            "diagram"
        } else if (rect is Decal) {
            "decal"
        } else if (rect is ConnectedRectangular) {
            "connected"
        } else if (rect is Label) {
            "label"
        } else if (rect is Terminator) {
            "terminator"
        } else {
            "unknown"
        }
    }

    /**
     * Outputs any SVG-renderable temporary elements to the output.
     */
    protected fun handleTemporaryElements(out: Element, d: Document, postProcessor: XMLProcessor) {
        if (r is Container) {
            (r as Container).getContents()
                .filterIsInstance<Temporary>()
                .filterIsInstance<HasSVGRepresentation>()
                .forEach {
                    val e = it.output(d, postProcessor)
                    if (e != null) {
                        e.setAttribute("k9-elem", "_temporary")
                        e.setAttribute("id", it.getID())
                        out.appendChild(e)
                    }
                }
        }
    }
}